pyrtl/passes.py

```diff
@@ -52,11 +52,100 @@ def optimize(update_working_block=True, block=None, skip_sanity_check=False):
+
+    # wire_creator maps from WireVector to the LogicNet that defines its value.
+    # wire_users maps from WireVector to a list of LogicNets that use its value.
+    wire_creator, wire_users = block.net_connections()
+
+    # Build a list of inverter chains. Each inverter chain is a list of WireVectors,
+    # from source to destination.
+    inverter_chains = []
+    for current_dest, current_creator in wire_creator.items():
+        if current_creator.op != "~":
+            # Skip non-inverters.
+            continue
+
+        # The current inverter connects current_arg (a WireVector) to current_dest (also
+        # a WireVector).
+        current_arg = current_creator.args[0]
+        # current_users is the number of LogicNets that use current_dest.
+        current_users = len(wire_users[current_dest])
+
+        # Add the current inverter to the end of this inverter chain.
+        append_to = None
+        # Add the current inverter to the beginning of this inverter chain.
+        prepend_to = None
+        next_inverter_chains = []
+        for inverter_chain in inverter_chains:
+            chain_arg = inverter_chain[0]
+            chain_dest = inverter_chain[-1]
+            chain_users = len(wire_users[chain_dest])
+
+            if chain_dest is current_arg and chain_users == 1:
+                # This chain's only destination is the current inverter. Append the
+                # current inverter to the chain.
+                append_to = inverter_chain
+            elif chain_arg is current_dest and current_users <= 1:
+                # This chain's only argument is the current inverter. Add the current
+                # inverter to the beginning of the chain.
+                prepend_to = inverter_chain
+            else:
+                next_inverter_chains.append(inverter_chain)
+
+        if append_to and prepend_to:
+            # The current inverter joins two existing inverter chains.
+            next_inverter_chains.append(append_to + prepend_to)
+        elif append_to:
+            # Add the current inverter after 'append_to'.
+            next_inverter_chains.append(append_to + [current_dest])
+        elif prepend_to:
+            # Add the current inverter before 'prepend_to'.
+            next_inverter_chains.append([current_arg] + prepend_to)
+        else:
+            next_inverter_chains.append([current_arg, current_dest])
+
+        inverter_chains = next_inverter_chains
+
+    new_logic = set()
+    net_removal_set = set()
+    wire_removal_set = set()
+    for inverter_chain in inverter_chains:
+        if len(inverter_chain) > 2:
+            if len(inverter_chain) % 2 == 1:  # even number of inverters in a chain
+                end_idx = len(inverter_chain) - 1
+            else:  # odd number of inverters in a chain
+                end_idx = len(inverter_chain) - 2
+            wires_to_remove = inverter_chain[1:end_idx]
+            new_logic.add(LogicNet('w', None, args=(inverter_chain[0],),
```

Comment:
Try writing this pass so we don't create new wires, then immediately remove them on line 146 below. It seems possible to rewrite the block, updating any LogicNets that use the end of an inverter chain to instead use the beginning of that inverter chain. The code should be similar to the second pass in `_remove_wire_nets` [here](https://github.com/UCSBarchlab/PyRTL/blob/development/pyrtl/passes.py#L92)